### ISSUE TITLE
1.3.20 breaks certain Alembic patterns with ArgumentError: column object already assigned to table

### ISSUE THREAD

**Issue by dtantsur** [sqlalchemy/sqlalchemy#5669]
**Describe the bug**
The ironic-inspector project is using two instances of a similar pattern in its alembic migrations: add a column, then load a table, then do some data manipulations. [Here is a concise example](https://opendev.org/openstack/ironic-inspector/src/commit/2e6f5ed25b738d186981ad21d13ecc1fbf2828e4/ironic_inspector/migrations/versions/d00d6e3f38c4_change_created_finished_at_type.py#L43-L47). This has worked up to and including 1.3.19 and fails in 1.3.20 with ArgumentError.

**Expected behavior**
I would be happy if it continued to work. Maybe we're doing something exceptionally stupid? Then some guidance is welcome. We are working around the problem by [switching to reflection for Table](https://review.opendev.org/#/c/759420/).

**To Reproduce**
You can do it using the ironic-inspector code tree:
```
pip3 install --user tox
git clone https://opendev.org/openstack/ironic-inspector
cd ironic-inspector
git checkout 2e6f5ed25b738d186981ad21d13ecc1fbf2828e4 # before the workaround
tox -epy3 -- migrations
```
Downgrade SQLAlchemy to see it working again:
```
./.tox/py3/bin/pip install 'SQLAlchemy===1.3.19'
tox -epy3 -- migrations
```

**Error**

```
2020-10-23 12:28:44.148029 | ubuntu-focal |         "  File \"/opt/stack/bifrost/lib/python3.8/site-packages/ironic_inspector/migrations/versions/d00d6e3f38c4_change_created_finished_at_type.py\", line 46, in upgrade",
2020-10-23 12:28:44.148057 | ubuntu-focal |         "    t = sa.table('nodes', started_at, finished_at,",
2020-10-23 12:28:44.148094 | ubuntu-focal |         "  File \"<string>\", line 2, in table",
2020-10-23 12:28:44.148121 | ubuntu-focal |         "  File \"/opt/stack/bifrost/lib/python3.8/site-packages/sqlalchemy/sql/selectable.py\", line 1963, in __init__",
2020-10-23 12:28:44.148149 | ubuntu-focal |         "    self.append_column(c)",
2020-10-23 12:28:44.148210 | ubuntu-focal |         "  File \"/opt/stack/bifrost/lib/python3.8/site-packages/sqlalchemy/sql/selectable.py\", line 1984, in append_column",
2020-10-23 12:28:44.148243 | ubuntu-focal |         "    raise exc.ArgumentError(",
2020-10-23 12:28:44.148278 | ubuntu-focal |         "sqlalchemy.exc.ArgumentError: column object 'temp_started_at' already assigned to table 'nodes'"
```

**Versions.**
 - OS: Fedora 32, Ubuntu Focal
 - Python: 3.8.6, likely other.
 - SQLAlchemy: 1.3.20
 - Database: N/A
 - DBAPI: N/A

**Have a nice day!**

**Comment by zzzeek** [comment]
Hi Dmitry -

first off I'm not sure if you're aware, I work on Openstack all day and this is an ironic-inspector bug, so normally what you can do here is make the launchpad bug and then find me on openstack freenode and I can help propose a new gerrit to fix.

the error here is that ironic-inspector is relying upon previously undefined and quite broken behavior which is that column objects allowed themselves to be attached to multiple tables, causing them to behave incorrectly.   ironic-inspector will need to modify its code to ensure that each Column object used only against a single table or operation; the alembic op.add_column() operation internally attaches the given column to a table object in order to render the appropriate SQL.  so these columns can't be repurposed after the fact.

**Comment by zzzeek** [comment]
also at *best* this would be an Alembic bug, where alembic could perhaps copy the column before using it for op.add_column().

**Comment by dtantsur** [comment]
@zzzeek I have indeed entirely forgotten that you work on openstack. Busy time..

Thank you for your explanation, I've applied it to https://review.opendev.org/759420.

**Comment by Adrian-Turjak** [comment]
@zzzeek Hey! We're actually hitting the same error in Gnocchi, but it's kinda unclear as to what the actual cause is, because we don't seem to be reusing things as far as I can find in the code.

This is the issue:
https://github.com/gnocchixyz/gnocchi/issues/1080

And the two migrations that are throw errors are:
`sqlalchemy.exc.ArgumentError: column object 'started_at_ts' already assigned to table 'resource'` at: https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/5c4f93e5bb4_mysql_float_to_timestamp.py
and
`sqlalchemy.exc.ArgumentError: column object 'creator' already assigned to table 'resource'` at https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/aba5a217ca9b_merge_created_in_creator.py

**Comment by zzzeek** [comment]
> @zzzeek Hey! We're actually hitting the same error in Gnocchi, but it's kinda unclear as to what the actual cause is, because we don't seem to be reusing things as far as I can find in the code.
> 
> This is the issue:
> [gnocchixyz/gnocchi#1080](https://github.com/gnocchixyz/gnocchi/issues/1080)
> 
> And the two migrations that are throw errors are:
> `sqlalchemy.exc.ArgumentError: column object 'started_at_ts' already assigned to table 'resource'` at: https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/5c4f93e5bb4_mysql_float_to_timestamp.py

so just look for combinations of table() and add_column(), for that one it's this:

https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/5c4f93e5bb4_mysql_float_to_timestamp.py#L66

```
    op.add_column(table_name, temp_col)
    t = sa.sql.table(table_name, existing_col, temp_col)

```
use a lambda for temp_col:

```
temp_col = lambda: sa.Column(
                column_name + "_ts",
                sqlalchemy_types.TimestampUTC(),
                nullable=True)
            op.add_column(table_name, temp_col())
            t = sa.sql.table(table_name, existing_col, temp_col())

```

> and
> `sqlalchemy.exc.ArgumentError: column object 'creator' already assigned to table 'resource'` at https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/aba5a217ca9b_merge_created_in_creator.py


right here:

https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/alembic/versions/aba5a217ca9b_merge_created_in_creator.py#L42


basically anytime you are assigning a Column to a variable:

```
  my_col = Column(...)
  op.add_column(..., my_col)
  t = table(..., my_col)

```
change to a lambda:

```
  my_col = lambda: Column(...)
  op.add_column(..., my_col())
  t = table(..., my_col())

```

**Comment by zzzeek** [comment]
we're going to fix this on the Alembic side as well as it's hitting me now too.   see https://github.com/sqlalchemy/alembic/issues/753
